File: examples/bridges/python_peer/publish.py
#!/usr/bin/env python3
"""Publish frames as the ``python_tooling`` peer.

Default args target the x86_dev UDS profile:

    examples/bridges/python_peer/publish.py --count 5

Each frame:
    source       = 7  (router rewrites this on forward; informational here)
    topic_id     = --topic-id (default 0)
    seq          = monotonic 0..count-1
    timestamp_ns = router_now_ns() (CLOCK_MONOTONIC_RAW per ADR 0010)
    payload      = --payload, suffixed with seq

Per the x86_dev.toml route ``source = 7 dest = [3]`` the recorder
receives every frame; downstream peers can be added by extending the
route table within the 2-destination cap (parked review C5).
"""

from __future__ import annotations

import argparse
import signal
import sys
import time

from rim_router_peer import RouterPeer, router_now_ns


def parse_args() -> argparse.Namespace:
    p = argparse.ArgumentParser(description=__doc__)
    p.add_argument("--router-path", default="/tmp/rim_router.sock",
                   help="router listen path (default: %(default)s)")
    p.add_argument("--peer-path", default="/tmp/rim_router_python_tooling.sock",
                   help="this peer's bound path (default: %(default)s)")
    p.add_argument("--peer-id", type=int, default=7,
                   help="this peer's id (default: %(default)s)")
    p.add_argument("--count", type=int, default=5,
                   help="number of frames to publish")
    p.add_argument("--interval-ms", type=int, default=80,
                   help="sleep between publishes (ms)")
    p.add_argument("--payload", default="python-tooling",
                   help="payload prefix; gets '-<seq>' appended")
    p.add_argument("--topic-id", type=int, default=0,
                   help="topic id stamped into every frame")
    return p.parse_args()


def main() -> int:
    args = parse_args()

    stop = {"flag": False}

    def _stop(*_: object) -> None:
        stop["flag"] = True

    signal.signal(signal.SIGINT, _stop)
    signal.signal(signal.SIGTERM, _stop)

    with RouterPeer(
        router_path=args.router_path,
        peer_path=args.peer_path,
        peer_id=args.peer_id,
    ) as peer:
        print(
            f"[publish] bound {args.peer_path} -> router {args.router_path}; "
            f"peer_id={args.peer_id}; sending {args.count} frame(s)",
            file=sys.stderr,
        )

        for seq in range(args.count):
            if stop["flag"]:
                break
            payload = f"{args.payload}-{seq}".encode("utf-8")
            frame = peer.publish(
                payload=payload,
                topic_id=args.topic_id,
                seq=seq,
                timestamp_ns=router_now_ns(),
            )
            print(
                f"[send] seq={frame.seq} topic={frame.topic_id} "
                f"ts_ns={frame.timestamp_ns} payload={frame.payload_bytes()!r}"
            )
            if seq < args.count - 1 and args.interval_ms > 0:
                time.sleep(args.interval_ms / 1000.0)

    print(f"[publish] done; sent {args.count} frame(s)", file=sys.stderr)
    return 0


if __name__ == "__main__":
    sys.exit(main())

File: examples/bridges/node_gateway/publish.js
#!/usr/bin/env node
// Publish N frames as a UDP peer, useful for driving smoke tests and for
// manual exercise of the gateway. Mirrors python_peer/publish.py.
//
// Defaults target the hil.toml UDP profile's sensor peer (id=1). The
// router resolves the source byte from the sender's bound UDP port (see
// peer_id_from_recv<Udp>), so we must bind at the sensor port before
// sending — handled by RouterPeer.open().

import { parseArgs } from 'node:util';

import { RouterPeer, routerNowNs } from './rim_router_peer.js';

function parseCli() {
    const { values } = parseArgs({
        options: {
            'router-host': { type: 'string', default: '127.0.0.1' },
            'router-port': { type: 'string', default: '19100' },
            'peer-host':   { type: 'string', default: '127.0.0.1' },
            'peer-port':   { type: 'string', default: '19101' },
            'peer-id':     { type: 'string', default: '1' },
            'count':       { type: 'string', default: '5' },
            'interval-ms': { type: 'string', default: '30' },
            'topic-id':    { type: 'string', default: '0' },
            'payload':     { type: 'string', default: 'node-frame' },
            'help':        { type: 'boolean', default: false },
        },
    });
    if (values.help) {
        process.stderr.write(
            'usage: publish.js [--router-host H] [--router-port P]\n' +
            '                  [--peer-host H]   [--peer-port P]   [--peer-id N]\n' +
            '                  [--count N]       [--interval-ms MS]\n' +
            '                  [--topic-id N]    [--payload TEXT]\n' +
            '\n' +
            'Each published frame gets a payload of "<payload>-<seq>" (e.g. node-frame-0,\n' +
            'node-frame-1, ...). seq counts up from 0.\n'
        );
        process.exit(0);
    }
    return values;
}

function sleep(ms) {
    return new Promise((resolve) => setTimeout(resolve, ms));
}

async function main() {
    const args = parseCli();
    const count = Number(args.count);
    const interval = Number(args['interval-ms']);
    const topicId = Number(args['topic-id']);
    const basePayload = args.payload;

    const peer = new RouterPeer({
        routerHost: args['router-host'],
        routerPort: args['router-port'],
        peerHost:   args['peer-host'],
        peerPort:   args['peer-port'],
        peerId:     args['peer-id'],
    });

    await peer.open();
    process.stderr.write(
        `[publish] bound udp ${args['peer-host']}:${args['peer-port']} ` +
        `as peer_id=${args['peer-id']} -> router ${args['router-host']}:${args['router-port']}; ` +
        `count=${count} interval_ms=${interval}\n`
    );

    for (let seq = 0; seq < count; seq += 1) {
        const payload = `${basePayload}-${seq}`;
        await peer.publish({
            payload,
            topic_id: topicId,
            seq,
            timestamp_ns: routerNowNs(),
        });
        process.stdout.write(
            `[publish] seq=${seq} payload=${JSON.stringify(payload)}\n`
        );
        if (seq < count - 1 && interval > 0) {
            await sleep(interval);
        }
    }

    await peer.close();
    process.stderr.write(`[publish] done; sent ${count} frame(s)\n`);
}

main().catch((err) => {
    process.stderr.write(`[publish] fatal: ${err.message}\n`);
    process.exit(1);
});
